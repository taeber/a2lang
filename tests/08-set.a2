; Exhaustive tests for assignments.

asm {
	ORG $800
	JSR main
	JMP EXIT
}

use [
    EXIT  : sub @ $3D0
    INIT  : sub @ $FB2F
    HOME  : sub @ $FC58
    COUT  : sub <- [ch: char @ A] @ $FDED
    CROUT : sub @ $FD8E
    PRBYTE: sub <- [byte: byte @ A] @ $FDDA
]

let Assert = sub <- [actual: byte, expected: byte] {
    PRBYTE(expected)
    if actual == expected {
        COUT(`=)
        COUT(`=)
        PRBYTE(actual)
        CROUT()
        ->
    }
    COUT(`<)
    COUT(`>)
    PRBYTE(actual)
    CROUT()
    COUT(`F)
    COUT(`A)
    COUT(`I)
    COUT(`L)
    COUT(`!)
    CROUT()
    EXIT()
}

let (
    PTR1 = $06
    PTR2 = $08
    PTR3 = $FE
)

let (
    NUL = $00
    ONE = 1
    TWO = 2
)

let Dimensions = [Width: int, Height: int, Depth: int]

var [
    varb: byte
    varw: word
    dims: Dimensions

    arrb: byte^12
    arrw: word^6
    arrg: Dimensions^4

    ptrb: byte^ @ PTR1
    ptrw: word^ @ PTR2
    ptrg: Dimensions^ @ PTR3

    rega: byte @ A
    regx: byte @ X
    regy: byte @ Y
    regya: word @ YA
]

let OutputOne = sub -> [one: int] {
    one := 1
    ->
}

let OutputRegA = sub -> [acc: byte @ A] {
    acc := `A
    ->
}

let main = sub {
    varb := OutputRegA()
    varb := OutputOne()

    rega := 5
    regx := 5
    regy := 5
    regy := regx
    varb := $9021
    regya := ptrb_6

    regya := $9021

    varb := $42      ; SETBB varb #$42
    varb := $9021    ; SETBB varb #$21

    varw := $42      ; SETWB varw #$42
    varw := $9021    ; SETWW varw #$9021

    ;dims := (Width=1, Height=2, Depth=3)
    dims.Width := 1
    dims.Height := 2
    dims.Depth := 3

    arrb_2   := varb; SETBB arrb+2 varb
    arrb_TWO := varw; SETBB arrb+TWO varw
    arrb_varb := 42 ; SETBB arrb,varb #42
    ; arrb_varw := 100  ; error: variable index is not byte size: varw.

    ptrb_6 := $21
    ; ptrw_$03 := $21    ; error: only byte pointers can be indexed: ptrw_$03

    ptrg := dims
    ptrg.Depth := 100
}

