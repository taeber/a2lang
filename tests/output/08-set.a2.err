warning: right-hand side will be truncated to a byte.
warning: right-hand side will be truncated to a byte.
warning: right-hand side will be truncated to a byte.
