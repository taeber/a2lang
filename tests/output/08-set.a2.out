	ORG $800
	JSR main
	JMP EXIT
EXIT	EQU $3D0
INIT	EQU $FB2F
HOME	EQU $FC58
COUT	EQU $FDED
CROUT	EQU $FD8E
PRBYTE	EQU $FDDA
* COPYBB @A Assert.expected
Assert	LDA Assert.expected
	JSR PRBYTE
* IFEQ Assert.actual Assert.expected
*   Assert._1 Assert._2
Assert._0	LDA Assert.actual
	CMP Assert.expected
	BEQ Assert._1
	JMP Assert._2
* COPYBB @A #"="
Assert._1	LDA #"="
	JSR COUT
* COPYBB @A #"="
	LDA #"="
	JSR COUT
* COPYBB @A Assert.actual
	LDA Assert.actual
	JSR PRBYTE
	JMP CROUT
* COPYBB @A #"<"
Assert._2	LDA #"<"
	JSR COUT
* COPYBB @A #">"
	LDA #">"
	JSR COUT
* COPYBB @A Assert.actual
	LDA Assert.actual
	JSR PRBYTE
	JSR CROUT
* COPYBB @A #"F"
	LDA #"F"
	JSR COUT
* COPYBB @A #"A"
	LDA #"A"
	JSR COUT
* COPYBB @A #"I"
	LDA #"I"
	JSR COUT
* COPYBB @A #"L"
	LDA #"L"
	JSR COUT
* COPYBB @A #"!"
	LDA #"!"
	JSR COUT
	JSR CROUT
	JMP EXIT
PTR1	EQU $06
PTR2	EQU $08
PTR3	EQU $FE
NUL	EQU $00
ONE	EQU $01
TWO	EQU $02
ptrb	EQU PTR1
ptrw	EQU PTR2
ptrg	EQU PTR3
* COPYBB OutputOne.one #$01
OutputOne	LDA #$01
	STA OutputOne.one
	RTS
* COPYBB @A #"A"
OutputRegA	LDA #"A"
	RTS
main	JSR OutputRegA
* COPYBB varb @A
	STA varb
	JSR OutputOne
* COPYBB varb OutputOne.one
	LDA OutputOne.one
	STA varb
* COPYBB @A #$05
	LDA #$05
* COPYBB @X #$05
	LDX #$05
* COPYBB @Y #$05
	LDY #$05
* COPYBB @Y @X
	TXA
	TAY
* COPYBB varb #$90,#$21
	LDA #$21
	STA varb
* WARNING: VALUE TRUNCATED
* COPYWB @YA (ptrb),#$06
	LDY #$06
	LDA (ptrb),Y
	LDY #0
* COPYWW @YA #$90,#$21
	LDA #$21
	LDY #$90
* COPYBB varb #$42
	LDA #$42
	STA varb
* COPYBB varb #$90,#$21
	LDA #$21
	STA varb
* WARNING: VALUE TRUNCATED
* COPYWB varw #$42
	LDA #$42
	LDX #0
	STA varw
	STX varw+1
* COPYWW varw #$90,#$21
	LDA #$21
	LDX #$90
	STA varw
	STX varw+1
* COPYBB dims+0 #$01
	LDA #$01
	STA dims+0
* COPYBB dims+1 #$02
	LDA #$02
	STA dims+1
* COPYBB dims+2 #$03
	LDA #$03
	STA dims+2
* COPYBB arrb+$02 varb
	LDA varb
	STA arrb+$02
* COPYBB arrb+$02 varw
	LDA varw
	STA arrb+$02
* WARNING: VALUE TRUNCATED
* COPYBB arrb,varb #$2A
	LDA #$2A
	LDY varb
	STA arrb,Y
* COPYBB (ptrb),#$06 #$21
	LDA #$21
	LDY #$06
	STA (ptrb),Y
	LDA #<dims
	LDX #>dims
	STX ptrg+1
	STA ptrg
* COPYBB (ptrg),#2 #$64
	LDA #$64
	LDY #2
	STA (ptrg),Y
	RTS
Assert.actual	HEX 00
Assert.expected	HEX 00
varb	HEX 00
varw	HEX 0000
dims	HEX 000000
arrb	HEX 000000000000000000000000
arrw	HEX 000000000000000000000000
arrg	HEX 000000000000000000000000
OutputOne.one	HEX 00
